File: head/src/main/kotlin/io/qalipsis/core/head/model/Campaign.kt
/*
 * QALIPSIS
 * Copyright (C) 2022 AERIS IT Solutions GmbH
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 *
 */

package io.qalipsis.core.head.model

import io.micronaut.core.annotation.Introspected
import io.qalipsis.api.report.ExecutionStatus
import io.swagger.v3.oas.annotations.media.Schema
import java.time.Instant
import javax.validation.constraints.PositiveOrZero

/**
 * External representation of a campaign.
 *
 * @author Eric Jessé
 */
@Introspected
@Schema(
    name = "Campaign details",
    title = "Details of a running or completed campaign"
)
open class Campaign(
    @field:Schema(description = "Last change of the campaign", required = true)
    val version: Instant,

    @field:Schema(description = "Unique identifier of the campaign", required = true)
    val key: String,

    @field:Schema(description = "Creation time of the campaign", required = true)
    val creation: Instant,

    @field:Schema(description = "Display name of the campaign", required = true)
    val name: String,

    @field:Schema(
        description = "Speed factor to apply on the execution profile, each strategy will apply it differently depending on its own implementation",
        required = true,
        example = "1.0"
    )
    val speedFactor: Double,

    @field:Schema(description = "Counts of minions scheduled to be started", required = false, example = "100")
    @field:PositiveOrZero
    val scheduledMinions: Int?,

    @field:Schema(
        description = "Instant when the campaign should be aborted without generating a failure",
        required = false
    )
    val softTimeout: Instant? = null,

    @field:Schema(
        description = "Instant when the campaign should be aborted as well as generating a failure",
        required = false
    )
    val hardTimeout: Instant? = null,

    @field:Schema(description = "Date and time when the campaign started", required = true)
    val start: Instant?,

    @field:Schema(
        description = "Date and time when the campaign was completed, whether successfully or not",
        required = false
    )
    val end: Instant?,

    @field:Schema(description = "Overall execution status of the campaign when completed", required = false)
    val status: ExecutionStatus,

    @field:Schema(description = "The root cause of the campaign failure", required = false)
    val failureReason: String? = null,

    @field:Schema(description = "Name of the user, who created the campaign", required = false, example = "John Doe")
    val configurerName: String?,

    @field:Schema(description = "Name of the user, who aborted the campaign", required = false, example = "John Doe")
    val aborterName: String? = null,

    @field:Schema(description = "Scenarios being part of the campaign", required = true)
    val scenarios: Collection<Scenario>,

    @field:Schema(description = "Keys of the zones where the campaign was executed", required = false)
    val zones: Set<String> = emptySet()
) {

    fun copy(
        version: Instant = this.version,
        key: String = this.key,
        creation: Instant = this.creation,
        name: String = this.name,
        speedFactor: Double = this.speedFactor,
        scheduledMinions: Int? = this.scheduledMinions,
        softTimeout: Instant? = this.softTimeout,
        hardTimeout: Instant? = this.hardTimeout,
        start: Instant? = this.start,
        end: Instant? = this.end,
        status: ExecutionStatus = this.status,
        failureReason: String? = null,
        configurerName: String? = this.configurerName,
        aborterName: String? = this.aborterName,
        scenarios: Collection<Scenario> = this.scenarios
    ) = Campaign(
        version = version,
        key = key,
        creation = creation,
        name = name,
        speedFactor = speedFactor,
        scheduledMinions = scheduledMinions,
        softTimeout = softTimeout,
        hardTimeout = hardTimeout,
        start = start,
        end = end,
        status = status,
        failureReason = failureReason,
        configurerName = configurerName,
        aborterName = aborterName,
        scenarios = scenarios
    )

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (javaClass != other?.javaClass) return false

        other as Campaign

        if (version != other.version) return false
        if (key != other.key) return false
        if (creation != other.creation) return false
        if (name != other.name) return false
        if (speedFactor != other.speedFactor) return false
        if (scheduledMinions != other.scheduledMinions) return false
        if (softTimeout != other.softTimeout) return false
        if (hardTimeout != other.hardTimeout) return false
        if (start != other.start) return false
        if (end != other.end) return false
        if (status != other.status) return false
        if (failureReason != other.failureReason) return false
        if (configurerName != other.configurerName) return false
        if (aborterName != other.aborterName) return false
        if (scenarios != other.scenarios) return false
        if (zones != other.zones) return false

        return true
    }

    override fun hashCode(): Int {
        var result = version.hashCode()
        result = 31 * result + key.hashCode()
        result = 31 * result + creation.hashCode()
        result = 31 * result + name.hashCode()
        result = 31 * result + speedFactor.hashCode()
        result = 31 * result + (scheduledMinions ?: 0)
        result = 31 * result + (softTimeout?.hashCode() ?: 0)
        result = 31 * result + (hardTimeout?.hashCode() ?: 0)
        result = 31 * result + (start?.hashCode() ?: 0)
        result = 31 * result + (end?.hashCode() ?: 0)
        result = 31 * result + status.hashCode()
        result = 31 * result + (configurerName?.hashCode() ?: 0)
        result = 31 * result + (aborterName?.hashCode() ?: 0)
        result = 31 * result + scenarios.hashCode()
        result = 31 * result + zones.hashCode()
        return result
    }

    override fun toString(): String {
        return "Campaign(version=$version, key='$key', creation=$creation, name='$name', speedFactor=$speedFactor, scheduledMinions=$scheduledMinions, softTimeout=$softTimeout, hardTimeout=$hardTimeout, start=$start, end=$end, result=$status, configurerName=$configurerName, aborterName=$aborterName, scenarios=$scenarios, zones=$zones)"
    }


}

File: utils/campaign.ts
export interface Campaign {
    /**
     * Last change of the campaign
     */
    version: string;

    /**
     * Unique identifier of the campaign
     */
    key: string;

    /**
     * Creation time of the campaign in ISO string format
     */
    creation: string;

    /**
     * Display name of the campaign
     */
    name: string;

    /**
     * Speed factor to apply on the execution profile, each strategy will apply it differently 
     * depending on its own implementation
     */
    speedFactor: number;

    /**
     * Counts of minions scheduled to be started
     */
    scheduledMinions?: number;

    /**
     * Instant when the campaign should be aborted without generating a failure in ISO string format
     */
    softTimeout?: string;

    /**
     * Instant when the campaign should be aborted as well as generating a failure
     */
    hardTimeout?: string;

    /**
     * Date and time when the campaign started
     */
    start?: string;

    /**
     * Date and time when the campaign was completed, whether successfully or not
     */
    end?: string;

    /**
     * Overall execution status of the campaign when completed
     */
    status?: ExecutionStatus;

    /**
     * The root cause of the campaign failure
     */
    failureReason?: string;

    /**
     * Name of the user, who created the campaign
     */
    configurerName?: string;

    /**
     * Name of the user, who aborted the campaign
     */
    aborterName?: string;

    /**
     * Scenarios being part of the campaign
     */
    scenarios: Scenario[];

    /**
     * Keys of the zones where the campaign was executed
     */
    zones?: string;
}

export interface CampaignTableData extends Campaign {
    /**
     * The elapsed time of the campaign.
     */
    elapsedTime: string;

    /**
     * Creation time of the campaign ih formatted text.
     */
    creationTime: string;

    /**
     * The text of all scenario names.
     */
    scenarioText: string;
    
    /**
     * The tag of the status.
     */
    statusTag: Tag;

    /**
     * A flag to indicate if the row can be selected.
     */
    disabled?: boolean;
}


/**
 * Execution status of a ScenarioReport or CampaignReport.
 */
export enum ExecutionStatus {
    /**
     * SUCCESSFUL - all the steps, were successful
     */
    SUCCESSFUL = 'SUCCESSFUL',

    /**
     * WARNING - a deeper look at the reports is required, but the campaign does not fail
     */
    WARNING = 'WARNING',

    /**
     * FAILED - the campaign went until the end, but got errors
     */
    FAILED = 'FAILED',

    /**
     * ABORTED - the campaign was aborted, either by a user or a critical failure
     */
    ABORTED = 'ABORTED',

    /**
     * SCHEDULED - the campaign is scheduled for a later point in time
     */
    SCHEDULED = 'SCHEDULED',

    /**
     * QUEUED - the campaign is being prepared and will start very soon
     */
    QUEUED = 'QUEUED',

    /**
     * IN_PROGRESS - the campaign is currently running
     */
    IN_PROGRESS = 'IN_PROGRESS'
}

/**
 * Details of the execution of a completed or running campaign and its scenario.
 */
export interface CampaignExecutionDetails {
    version: string;
    key: string;
    creation: string;
    name: string;
    speedFactor: number;
    scheduledMinions?: number;
    timeout?: string;
    hardTimeout?: boolean;
    start?: string;
    end?: string;
    status: ExecutionStatus;
    failureReason?: string;
    configurerName?: string;
    aborterName?: string;
    scenarios?: Scenario[];
    zones?: string[];
    /**
     * Counts of minions when the campaign started.
     */
    startedMinions: number;

    /**
     * Counts of minions that completed the campaign.
     */
    completedMinions: number;

    /**
     * Counts of steps that successfully completed.
     */
    successfulExecutions: number;

    /**
     * Counts of steps that failed.
     */
    failedExecutions: number;

    /**
     * Individual details of the scenario executed during the campaign.
     */
    scenariosReports: ScenarioExecutionDetails[];
}
